tunneld/options.go

```diff
@@ -99,59 +100,106 @@ func (options *Options) Validate() error {
 	if options.WireguardNetworkPrefix.Bits() > 64 {
 		return xerrors.New("WireguardNetworkPrefix must have at least 64 bits available")
 	}
+	if options.WireguardNetworkPrefix.Bits()%8 != 0 {
+		return xerrors.New("WireguardNetworkPrefix must be a multiple of 8 bits")
+	}
 	if !options.WireguardNetworkPrefix.Contains(options.WireguardServerIP) {
 		return xerrors.New("WireguardServerIP must be contained within WireguardNetworkPrefix")
 	}
 
 	return nil
 }
 
-// WireguardPublicKeyToIP returns the IP address that corresponds to the given
-// wireguard public key.
-func (options *Options) WireguardPublicKeyToIP(publicKey device.NoisePublicKey) netip.Addr {
+// WireguardPublicKeyToIPAndURLs returns the IP address that corresponds to the
+// given wireguard public key, as well as all accepted tunnel URLs for the key.
+//
+// We support an older 32 character format ("old format") and a newer 12
+// character format ("good format") which is preferred. The first URL returned
+// should be considered "preferred", and all other URLs are provided for
+// compatibility with older deployments only.
+//
+// "good format":
+//
+//	Take the first 8 bytes of the hash of the public key, and convert to
+//	base32.
+//
+// "old format":
+//
+//	Take the network prefix, and create a new address filling the last n bytes
+//	with the first n bytes of the hash of the public key. Then convert to hex.
+func (options *Options) WireguardPublicKeyToIPAndURLs(publicKey device.NoisePublicKey, version tunnelsdk.TunnelVersion) (netip.Addr, []*url.URL) {
 	var (
 		keyHash   = sha256.Sum256(publicKey[:])
 		addrBytes = options.WireguardNetworkPrefix.Addr().As16()
 	)
 
-	// Copy the first 8 bytes of the hash to the last 8 bytes of the address. We
-	// only use 8 bytes because we only store 8 bytes of information in the
-	// generated hostname.
-	//
-	// The prefix has at least 64 bits of space available, so we can safely
-	// overwrite the last 8 bytes.
+	// IPv6 address:
+	// For the IP address, we take the first 64 bits of the network prefix and
+	// the first 64 bits of the hash of the public key.
 	copy(addrBytes[8:], keyHash[:8])
 
-	return netip.AddrFrom16(addrBytes)
-}
+	// Good format:
+	goodFormatBytes := make([]byte, 8)
+	copy(goodFormatBytes, keyHash[:8])
+	goodFormat := newHostnameEncoder.EncodeToString(goodFormatBytes)
+	goodFormatURL := *options.BaseURL
+	goodFormatURL.Host = strings.ToLower(goodFormat) + "." + goodFormatURL.Host
+
+	// Old format:
+	oldFormatBytes := make([]byte, 16)
+	copy(oldFormatBytes, addrBytes[:])
+	prefixLenBytes := options.WireguardNetworkPrefix.Bits() / 8
+	copy(oldFormatBytes[prefixLenBytes:], keyHash[:16-prefixLenBytes])
+	oldFormat := hex.EncodeToString(oldFormatBytes)
+	oldFormatURL := *options.BaseURL
+	oldFormatURL.Host = strings.ToLower(oldFormat) + "." + oldFormatURL.Host
```

Comment:
This "old format" seems different than the previous (old old format), but I suppose this is ok?

Reply:
You mean it's different to the old format generated by the current wgtunnel? It should be the same and there's tests for it...

Reply:
Ah, maybe just the way it was constructed changed then? I thought it looked different but maybe I'm blind, I trust you. 😄

Reply:
I'll try it with a real current coder client post-merge and see what happens